### PR TITLE
Get related charms from charmstore apiv4

### PR DESCRIPTION
**Pull request by hatched**
The format of responded data from the charmstore apiv3 and v4 were completely different so the parsing implementation needed to be rewritten.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

Open the charm details page *Related Charms* tab, it should list out the related charm icons.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2409/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2410/
Test PASSed.

File: app/models/charm.js [hunk 1]
```diff
@@ -277,54 +248,117 @@ YUI.add('juju-charm-models', function(Y) {
...
+        charms = this._keepLatestRevision(charms);
+        collection[name] = charms;
+      }, this);
+      return collection;
+    },
 
-      allCharmsList.sort(function(charm1, charm2) {
-        return charm2.weight - charm1.weight;
+    /**
+      Splits charm list into a key value delimited list of charms.
```
Comment (makyo) [inline app/models/charm.js]: trivial: "key/value"

File: app/models/charm.js [hunk 2]
```diff
@@ -277,54 +248,117 @@ YUI.add('juju-charm-models', function(Y) {
...
+
+      @method _splitIntoCharmCollections
+      @param {Array} indexes The list of id's for the charm list. It's an
+        integer list.
+      @param {String} name The interface name that the charms are sorted under.
+      @param {Object} charmList The full requires or provides interfact ordered
```
Comment (makyo) [inline app/models/charm.js]: trivial: "interface"

**Comment by makyo** [comment]
This looks great, much simpler. :+1: on code, will QA now.

**Comment by makyo** [comment]
QA looks good on most charms, but not http://localhost:8888/precise/ubuntu-4/#related-charms because Object.keys is being called on a non-object, charm.js:278  :-1: until that's taken care of more gracefully.

**Comment by hatched** [comment]
Thanks for the review and QA @makyo this should be fixed now, you can try with `/precise/ceph-28` which is missing only the provides as well as `/precise/ubuntu-4` which has neither.

**Comment by makyo** [comment]
Much better, thanks! :+1:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2412/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks @hatched I think it looks all good :+1:

**Comment by hatched** [comment]
Thanks for the reviews and qa's! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**